src/Microsoft.OpenApi/Models/OpenApiDocument.cs

```diff
@@ -268,6 +431,56 @@ private static void WriteHostInfoV2(IOpenApiWriter writer, IList<OpenApiServer>
             // schemes
             writer.WriteOptionalCollection(OpenApiConstants.Schemes, schemes, (w, s) => w.WriteValue(s));
         }
+        
+        private static async Task WriteHostInfoV2Async(IOpenApiWriter writer, IList<OpenApiServer> servers)
+        {
+            if (servers == null || !servers.Any())
+            {
+                return;
+            }
+
+            // Arbitrarily choose the first server given that V2 only allows 
+            // one host, port, and base path.
+            var firstServer = servers.First();
+
+            // Divide the URL in the Url property into host and basePath required in OpenAPI V2
+            // The Url property cannotcontain path templating to be valid for V2 serialization.
+            var firstServerUrl = new Uri(firstServer.Url);
+
+            // host
+            await writer.WritePropertyAsync(
+                OpenApiConstants.Host,
+                firstServerUrl.GetComponents(UriComponents.Host | UriComponents.Port, UriFormat.SafeUnescaped));
+
+            // basePath
+            if (firstServerUrl.AbsolutePath != "/")
+            {
+                await writer.WritePropertyAsync(OpenApiConstants.BasePath, firstServerUrl.AbsolutePath);
+            }
+
+            // Consider all schemes of the URLs in the server list that have the same
+            // host, port, and base path as the first server.
+            var schemes = servers.Select(
+                    s =>
+                    {
+                        Uri.TryCreate(s.Url, UriKind.RelativeOrAbsolute, out var url);
+                        return url;
+                    })
+                .Where(
+                    u => Uri.Compare(
+                            u,
+                            firstServerUrl,
+                            UriComponents.Host | UriComponents.Port | UriComponents.Path,
+                            UriFormat.SafeUnescaped,
+                            StringComparison.OrdinalIgnoreCase) ==
+                        0)
+                .Select(u => u.Scheme)
+                .Distinct()
+                .ToList();
```

Comment:
This section does not access `writer` and does not `await` anything. Could refactor it to a separate method, to reduce duplication of code.